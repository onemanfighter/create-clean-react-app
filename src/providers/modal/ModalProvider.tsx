import { useCallback, useEffect } from 'react';
import { useDisclosure } from '@chakra-ui/react';
import { ModalComponent } from '@components';
import {
  ModalOpenState,
  appStore,
  modalDataSelector,
  useShallow,
} from '@uiStore';

import { ModalProviderProps } from './types';

const ModalProvider = ({ children }: ModalProviderProps) => {
  const { isOpen, onOpen, onClose } = useDisclosure();
  const { openState, modalID, onModalClose, resetModalState } = appStore(
    useShallow(modalDataSelector),
  );

  const onModalCloseHandler = useCallback(() => {
    onModalClose && onModalClose();
    onClose();
    resetModalState();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [onModalClose]);

  useEffect(() => {
    if (openState === ModalOpenState.OPEN) {
      onOpen();
    } else {
      onModalCloseHandler();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [openState]);

  return (
    <>
      <ModalComponent
        modalID={modalID}
        onModalClose={onModalCloseHandler}
        isOpen={isOpen}
      />
      {children}
    </>
  );
};

export default ModalProvider;
