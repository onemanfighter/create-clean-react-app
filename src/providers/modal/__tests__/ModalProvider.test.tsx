import { render, screen } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import ModalProvider from '../ModalProvider';
import { appStore } from '@uiStore';

describe('ModalProvider', () => {
  it('should render correctly', () => {
    const { result } = renderHook(() => appStore());

    result.current.Modal.resetModalState();
    const { container } = render(<ModalProvider>App</ModalProvider>);

    jest.runAllTimersAsync();

    expect(container).toMatchSnapshot();
    expect(screen.getByText('App')).toBeDefined();
  });

  it('should render correctly with children', () => {
    const { result } = renderHook(() => appStore());

    result.current.Modal.resetModalState();
    const { container } = render(
      <ModalProvider>
        <div>App</div>
      </ModalProvider>,
    );

    jest.runAllTimersAsync();

    expect(container).toMatchSnapshot();
    expect(screen.getByText('App')).toBeDefined();
  });
});
