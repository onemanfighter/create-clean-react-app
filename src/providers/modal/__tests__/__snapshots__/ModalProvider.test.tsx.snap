// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalProvider should render correctly 1`] = `
<div>
  App
</div>
`;

exports[`ModalProvider should render correctly with children 1`] = `
<div>
  <div>
    App
  </div>
</div>
`;
